keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `provisioner`: The provisioner field defined in the `BucketClass`.  Used by sidecars to filter Buckets.
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield cases, backing storage instance is Retained, and never deleted.
+   - _Delete_:  the bucket and its contents are destroyed.
+   - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed.
+1. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.  If `BucketClass.anonymousAccessMode` is set, the value is copied here.  Accepted values:
+   - `private`: Default, disallow uncredentialed access to the storage instance.
+   - `ro`: Read only, uncredentialed users are permitted *read* operations for objects within the storage instance.
+   - `rw`: Read/Write, uncredentialed users are permitted *read and write* operations for objects within the storage instance. 
+   - `wo`: Write, uncredentialed users are permitted *write* operations for objects within the storage instance. 
+1. `bucketClassName`: Name of the associated BucketClass.
+1. `permittedNamespaces`: An array of namespaces, identified by a name and uid, from which  BucketRequests are allowed to bind to the Bucket.  Provided to allow admins a layer of cluster-layer access control.  Does **not** reflect or alter the backing storage instances' ACLs or IAM policies.
+   - In Greenfield, the originating BuckerRequest’s namespace must be specified at time of Bucket generation.
+   - In Brownfield, this list is defined by the admin.
+1. `protocol`: The protocol the application will use to access the storage instance.
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+1. `parameters`: a copy of the BucketClass parameters.
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Ready_: the controller finished processing and the controller is ready to receive access requests.
+   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketClass
+
```

Comment:
Is BucketClass _only_ used in greenfield?  @wlan0

Reply:
BucketClass is also used in brownfield requests. It is used to retrieve a bucket that satisfies the constraints laid out in the bucketClass

Reply:
Per conversation with @wlan0 : BC is only used for greenfield

Reply:
@jeffvance @wlan0  If BC is for greenfield only then for brownfield we still need `B.spec.allowedNamespaces` so admin who creates it can define which namespaces can bind to it. 

The only "confusion" we might want to clarify is in the greenfield case when BC.allowedNamespaces=[X,Y,Z] and namespace X creates a greenfield BR with this BC, then COSI provisions a new bucket, creates a new B, bind it to the BR, etc. But what does `B.allowedNamespaces` mean for greenfield Buckets? Is it a copy of the `BC.allowedNamespaces=[X,Y,Z]`? or is it restricted just to the namespace that provisioned it - B.allowedNamespaces=[X]? I think that copying from the BC would lead to a confusion - could namespace Y accidently/deliberately bind to namespace X greenfield bucket?

Reply:
> what does B.allowedNamespaces mean for greenfield Buckets? Is it a copy of the BC.allowedNamespaces=[X,Y,Z]?

It is a copy from the BC. The list can be mutated by the admin.

Reply:
@jeffvance I still feel like for greenfield it doesn't make any sense to copy allowedNamespaces from the BC into the B.

It's as if by sharing a BC between different users I am somehow allowing other namespaces to access my data.

Even if you somehow make sure the other namespaces cannot bind to this B, even declaratively this sounds like a bad idea to me. Is that concerning you as well?

Reply:
For greenfield, `BC.allowedNamespaces` gives the admin some control over who can create new buckets. Once a B is created based on a BR and the BC and bound to the BR, I agree that B.allowedNamespaces is not needed by COSI due to our 1:1 mapping of BR:B. And, B.allowedNamespaces could be set to the BR's ns and changed for brownfield after the admin clones this B.

For pure brownfield (not green -> brown), the admin creates a `Bucket` instance and defines which namespaces are allowed to request access to the B.  Here, the user BR does not need to reference a BC since the B already exists. But the KEP has gone back and forth on this. B.allowedNamespaces gives the admin control over who can request access to the B. Eg. a cluster scoped RBAC could allow all users to see Bs  but admin still wants to control who can use a B.

For the green->brownfield case, if B.allowedNamespaces contains only the ns of the BR which caused this B to be instantiated, then the admin needs to change this field after the copy.  This can work fine since the admin might also need to edit B.allowedNamespaces even if this value is based on BC.allowedNamespaces.